File: sdks/typescript/tests/webauthn.test.ts
/**
 * Unit tests for WebAuthn passkey ceremony helpers on HearthApiClient (C-21).
 * Tests are written before implementation (TDD).
 *
 * The browser SDK is the natural home for WebAuthn: the begin/complete calls
 * here return/consume the options that `navigator.credentials.create()` and
 * `navigator.credentials.get()` produce. These tests mock `fetch` and assert
 * the wire contract (endpoint, method, bearer header, body shape) matches the
 * Go/Python/Rust SDKs.
 */

import { describe, it, expect, vi, afterEach } from "vitest";
import { HearthApiClient } from "../src/client.js";

function makeClient(): HearthApiClient {
  return new HearthApiClient({
    baseUrl: "https://auth.example.com",
    realmId: "realm_test",
  });
}

function mockFetch(body: unknown, status = 200) {
  return vi
    .spyOn(globalThis, "fetch")
    .mockResolvedValue(new Response(JSON.stringify(body), { status }));
}

function lastCall(): [string, RequestInit] {
  return vi.mocked(fetch).mock.calls[0] as unknown as [string, RequestInit];
}

describe("HearthApiClient WebAuthn helpers (C-21)", () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  // ── Registration begin ──────────────────────────────────────────────────
  it("startWebAuthnRegistration POSTs to /webauthn/register/begin with bearer", async () => {
    const begin = {
      challenge: "chal-1",
      rp_id: "auth.example.com",
      rp_name: "Hearth",
      user_id: "user_1",
      user_name: "alice",
      user_display_name: "Alice",
      attestation: "none",
      timeout: 60000,
    };
    mockFetch(begin);
    const client = makeClient();

    const res = await client.startWebAuthnRegistration("bearer-token");

    const [url, init] = lastCall();
    expect(url).toBe("https://auth.example.com/webauthn/register/begin");
    expect(init.method).toBe("POST");
    expect((init.headers as Record<string, string>).Authorization).toBe(
      "Bearer bearer-token",
    );
    expect((init.headers as Record<string, string>)["X-Realm-ID"]).toBe("realm_test");
    expect(res.challenge).toBe("chal-1");
    expect(res.rp_id).toBe("auth.example.com");
  });

  // ── Registration complete ───────────────────────────────────────────────
  it("finishWebAuthnRegistration POSTs attestation with bearer and returns credential", async () => {
    mockFetch({ credential_id: "cred-1", algorithm: -8, discoverable: true });
    const client = makeClient();

    const res = await client.finishWebAuthnRegistration("bearer-token", {
      client_data_json: "cdj",
      attestation_object: "att",
      origin: "https://app.example.com",
      discoverable: true,
    });

    const [url, init] = lastCall();
    expect(url).toBe("https://auth.example.com/webauthn/register/complete");
    expect((init.headers as Record<string, string>).Authorization).toBe(
      "Bearer bearer-token",
    );
    const body = JSON.parse(init.body as string);
    expect(body.client_data_json).toBe("cdj");
    expect(body.attestation_object).toBe("att");
    expect(res.credential_id).toBe("cred-1");
  });

  // ── Authentication begin (discoverable / resident key) ──────────────────
  it("startWebAuthnAuthentication POSTs to /webauthn/auth/begin without bearer when no userId", async () => {
    mockFetch({
      challenge: "chal-2",
      rp_id: "auth.example.com",
      allow_credentials: [],
      user_verification: "preferred",
      timeout: 60000,
    });
    const client = makeClient();

    const res = await client.startWebAuthnAuthentication();

    const [url, init] = lastCall();
    expect(url).toBe("https://auth.example.com/webauthn/auth/begin");
    expect(init.method).toBe("POST");
    expect((init.headers as Record<string, string>).Authorization).toBeUndefined();
    const body = JSON.parse(init.body as string);
    expect(body.user_id).toBeUndefined();
    expect(res.challenge).toBe("chal-2");
  });

  it("startWebAuthnAuthentication includes user_id when provided", async () => {
    mockFetch({
      challenge: "chal-3",
      rp_id: "auth.example.com",
      allow_credentials: [{ id: "cred-1", type: "public-key" }],
      user_verification: "required",
      timeout: 60000,
    });
    const client = makeClient();

    await client.startWebAuthnAuthentication("user_42");

    const [, init] = lastCall();
    const body = JSON.parse(init.body as string);
    expect(body.user_id).toBe("user_42");
  });

  // ── Authentication complete → tokens ────────────────────────────────────
  it("finishWebAuthnAuthentication POSTs assertion and returns a TokenResponse", async () => {
    mockFetch({
      access_token: "at",
      id_token: "it",
      token_type: "Bearer",
      expires_in: 3600,
      refresh_token: "rt",
    });
    const client = makeClient();

    const res = await client.finishWebAuthnAuthentication({
      credential_id: "cred-1",
      client_data_json: "cdj",
      authenticator_data: "ad",
      signature: "sig",
      origin: "https://app.example.com",
    });

    const [url, init] = lastCall();
    expect(url).toBe("https://auth.example.com/webauthn/auth/complete");
    expect(init.method).toBe("POST");
    const body = JSON.parse(init.body as string);
    expect(body.credential_id).toBe("cred-1");
    expect(body.signature).toBe("sig");
    expect(res.access_token).toBe("at");
    expect(res.token_type).toBe("Bearer");
  });
});

File: sdks/kotlin/hearth-core/src/test/kotlin/io/hearth/sdk/WebAuthnTest.kt
package io.hearth.sdk

import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class WebAuthnTest {

    private lateinit var server: MockWebServer

    @BeforeTest
    fun setUp() {
        server = MockWebServer()
        server.start()
    }

    @AfterTest
    fun tearDown() {
        server.shutdown()
    }

    private fun makeClient(realmId: String = "realm-1") = HearthClient(
        issuerUrl = server.url("/").toString().trimEnd('/'),
        realmId = realmId,
    )

    // ── Registration begin ────────────────────────────────────────────────────

    @Test
    fun `startWebAuthnRegistration posts to webauthn register begin`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"challenge":"abc123","rp_id":"example.com","rp_name":"Example",""" +
                    """"user_id":"u1","user_name":"alice","user_display_name":"Alice",""" +
                    """"attestation":"none","timeout":60000}"""
                )
                .setResponseCode(200)
        )

        makeClient().startWebAuthnRegistration("bearer-token")

        val req = server.takeRequest()
        assertEquals("POST", req.method)
        assertTrue(req.path!!.endsWith("/webauthn/register/begin"), "path was: ${req.path}")
    }

    @Test
    fun `startWebAuthnRegistration sends Authorization header`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"challenge":"c","rp_id":"x","rp_name":"X","user_id":"u",""" +
                    """"user_name":"u","user_display_name":"U","attestation":"none","timeout":60000}"""
                )
                .setResponseCode(200)
        )

        makeClient().startWebAuthnRegistration("my-access-token")

        val req = server.takeRequest()
        assertEquals("Bearer my-access-token", req.getHeader("Authorization"))
    }

    @Test
    fun `startWebAuthnRegistration returns parsed response`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"challenge":"abc123","rp_id":"example.com","rp_name":"Example App",""" +
                    """"user_id":"user-1","user_name":"alice","user_display_name":"Alice Smith",""" +
                    """"attestation":"none","timeout":60000}"""
                )
                .setResponseCode(200)
        )

        val resp = makeClient().startWebAuthnRegistration("token")

        assertEquals("abc123", resp.challenge)
        assertEquals("example.com", resp.rpId)
        assertEquals("Example App", resp.rpName)
        assertEquals("user-1", resp.userId)
        assertEquals("alice", resp.userName)
        assertEquals("Alice Smith", resp.userDisplayName)
        assertEquals("none", resp.attestation)
        assertEquals(60_000L, resp.timeout)
    }

    // ── Registration finish ───────────────────────────────────────────────────

    @Test
    fun `finishWebAuthnRegistration posts to webauthn register complete`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody("""{"credential_id":"cred-abc","algorithm":-7,"discoverable":true}""")
                .setResponseCode(200)
        )

        makeClient().finishWebAuthnRegistration(
            "bearer-token",
            WebAuthnRegistrationCompleteRequest(
                clientDataJson = "eyJ...",
                attestationObject = "o2N...",
                origin = "https://example.com",
            ),
        )

        val req = server.takeRequest()
        assertEquals("POST", req.method)
        assertTrue(req.path!!.endsWith("/webauthn/register/complete"), "path was: ${req.path}")
    }

    @Test
    fun `finishWebAuthnRegistration returns parsed credential response`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody("""{"credential_id":"cred-xyz","algorithm":-8,"discoverable":false}""")
                .setResponseCode(200)
        )

        val resp = makeClient().finishWebAuthnRegistration(
            "token",
            WebAuthnRegistrationCompleteRequest(
                clientDataJson = "eyJ...",
                attestationObject = "o2N...",
                origin = "https://example.com",
                discoverable = true,
            ),
        )

        assertEquals("cred-xyz", resp.credentialId)
        assertEquals(-8L, resp.algorithm)
        assertEquals(false, resp.discoverable)
    }

    // ── Authentication begin ──────────────────────────────────────────────────

    @Test
    fun `startWebAuthnAuthentication posts to webauthn auth begin`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"challenge":"xyz789","rp_id":"example.com","allow_credentials":[],""" +
                    """"user_verification":"preferred","timeout":60000}"""
                )
                .setResponseCode(200)
        )

        makeClient().startWebAuthnAuthentication()

        val req = server.takeRequest()
        assertEquals("POST", req.method)
        assertTrue(req.path!!.endsWith("/webauthn/auth/begin"), "path was: ${req.path}")
    }

    @Test
    fun `startWebAuthnAuthentication without userId sends empty body`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"challenge":"c","rp_id":"x","allow_credentials":[],""" +
                    """"user_verification":"preferred","timeout":60000}"""
                )
                .setResponseCode(200)
        )

        makeClient().startWebAuthnAuthentication()

        val req = server.takeRequest()
        val body = req.body.readUtf8()
        assertTrue(!body.contains("user_id"), "body must not contain user_id when omitted: $body")
    }

    @Test
    fun `startWebAuthnAuthentication with userId includes user_id in body`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"challenge":"c","rp_id":"x","allow_credentials":[],""" +
                    """"user_verification":"preferred","timeout":60000}"""
                )
                .setResponseCode(200)
        )

        makeClient().startWebAuthnAuthentication(userId = "user-abc")

        val req = server.takeRequest()
        val body = req.body.readUtf8()
        assertTrue(body.contains("user-abc"), "body should contain userId: $body")
    }

    @Test
    fun `startWebAuthnAuthentication returns parsed response`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"challenge":"xyz789","rp_id":"example.com",""" +
                    """"allow_credentials":[{"id":"cred-1","type":"public-key"}],""" +
                    """"user_verification":"required","timeout":30000}"""
                )
                .setResponseCode(200)
        )

        val resp = makeClient().startWebAuthnAuthentication()

        assertEquals("xyz789", resp.challenge)
        assertEquals("example.com", resp.rpId)
        assertEquals(1, resp.allowCredentials.size)
        assertEquals("cred-1", resp.allowCredentials[0].id)
        assertEquals("public-key", resp.allowCredentials[0].type)
        assertEquals("required", resp.userVerification)
        assertEquals(30_000L, resp.timeout)
    }

    // ── Authentication finish ─────────────────────────────────────────────────

    @Test
    fun `finishWebAuthnAuthentication posts to webauthn auth complete`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"access_token":"eyJ...","token_type":"Bearer",""" +
                    """"expires_in":3600,"refresh_token":"ref..."}"""
                )
                .setResponseCode(200)
        )

        makeClient().finishWebAuthnAuthentication(
            WebAuthnAuthenticationCompleteRequest(
                credentialId = "cred-abc",
                clientDataJson = "eyJ...",
                authenticatorData = "SZYN...",
                signature = "abc...",
                origin = "https://example.com",
            ),
        )

        val req = server.takeRequest()
        assertEquals("POST", req.method)
        assertTrue(req.path!!.endsWith("/webauthn/auth/complete"), "path was: ${req.path}")
    }

    @Test
    fun `finishWebAuthnAuthentication returns TokenResponse`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"access_token":"eyJ.access","token_type":"Bearer",""" +
                    """"expires_in":3600,"refresh_token":"eyJ.refresh"}"""
                )
                .setResponseCode(200)
        )

        val resp = makeClient().finishWebAuthnAuthentication(
            WebAuthnAuthenticationCompleteRequest(
                credentialId = "cred-abc",
                clientDataJson = "eyJ...",
                authenticatorData = "SZYN...",
                signature = "abc...",
                origin = "https://example.com",
            ),
        )

        assertEquals("eyJ.access", resp.accessToken)
        assertEquals("Bearer", resp.tokenType)
        assertEquals(3600, resp.expiresIn)
        assertNotNull(resp.refreshToken)
    }

    @Test
    fun `finishWebAuthnAuthentication sends credential fields in body`() = runTest {
        server.enqueue(
            MockResponse()
                .setBody(
                    """{"access_token":"t","token_type":"Bearer","expires_in":3600}"""
                )
                .setResponseCode(200)
        )

        makeClient().finishWebAuthnAuthentication(
            WebAuthnAuthenticationCompleteRequest(
                credentialId = "cred-123",
                clientDataJson = "cdj-data",
                authenticatorData = "auth-data",
                signature = "sig-data",
                origin = "https://example.com",
                userHandle = "handle-abc",
            ),
        )

        val body = server.takeRequest().body.readUtf8()
        assertTrue(body.contains("cred-123"), "body should contain credential_id: $body")
        assertTrue(body.contains("cdj-data"), "body should contain client_data_json: $body")
        assertTrue(body.contains("handle-abc"), "body should contain user_handle: $body")
    }
}
